plugins/modem-manager/fu-mbim-qdu-updater.c

```diff
@@ -0,0 +1,464 @@
+		g_debug ("error: operation failed: %s\n", error->message);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (!mbim_message_qdu_update_session_response_parse (response,
+							     &out_current_session_type,
+							     &out_current_session_status,
+							     &out_last_session_type,
+							     &out_last_session_result,
+							     &out_last_session_error_offset,
+							     &out_last_session_error_size,
+							     &error)) {
+		g_debug ("error: couldn't parse response message: %s\n", error->message);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (response)
+		mbim_message_unref (response);
+
+	g_debug ("[%s] Successfully request modem to update session", mbim_device_get_path_display (device));
+
+	request = mbim_message_qdu_file_open_set_new (0, g_bytes_get_size (ctx->blob), NULL);
+	mbim_device_command (device,
+			     request,
+			     10,
+			     ctx->cancellable,
+			     (GAsyncReadyCallback)fu_mbim_qdu_updater_file_open_ready,
+			     ctx);
+	mbim_message_unref (request);
+}
+
+static void
+fu_mbim_qdu_updater_set_update_session (MbimDevice *device, WriteContext *ctx)
+{
+  	g_autoptr(MbimMessage) request = NULL;
+
+	request = (mbim_message_qdu_update_session_set_new (0, 1, NULL));
+	if (!request)
+		g_debug ("error: couldn't create request: \n");
+	else
+		g_debug ("Create request");
+
+	mbim_device_command (device,
+			     request,
+			     10,
+			     ctx->cancellable,
+			     (GAsyncReadyCallback)fu_mbim_qdu_updater_session_ready,
+			     ctx);
+	mbim_message_unref (request);
+}
+
+static GArray *
+fu_mbim_qdu_updater_get_checksum (GBytes *blob)
+{
+	gsize file_size;
+	gsize hash_size;
+	GArray *digest;
+	g_autoptr(GChecksum) checksum = NULL;
+
+	/* get checksum, to be used as unique id */
+	file_size = g_bytes_get_size (blob);
+	hash_size = g_checksum_type_get_length (G_CHECKSUM_SHA1);
+	checksum = g_checksum_new (G_CHECKSUM_SHA1);
+	g_checksum_update (checksum, g_bytes_get_data (blob, NULL), file_size);
+
+	/* libqmi expects a GArray of bytes, not a GByteArray */
+	digest = g_array_sized_new (FALSE, FALSE, sizeof (guint8), hash_size);
+	g_array_set_size (digest, hash_size);
+	g_checksum_get_digest (checksum, (guint8 *)digest->data, &hash_size);
+
+	return digest;
+}
+
+GArray *
+fu_mbim_qdu_updater_write (FuMbimQduUpdater *self, const gchar *filename, GBytes *blob, GError **error, FuDevice *device)
+{
+	g_autoptr(GMainLoop) mainloop = g_main_loop_new (NULL, FALSE);
+	g_autoptr(GArray) digest = fu_mbim_qdu_updater_get_checksum (blob);
+	WriteContext ctx = {
+		.mainloop = mainloop,
+		.mbim_device = self->mbim_device,
+		.cancellable = g_cancellable_new (),
+		.error = NULL,
+		.timeout_id = 0,
+		.blob = blob,
+		.digest = digest,
+		.chunks = NULL,
+		.chunk_sent = 0,
+		.device = device,
+	};
+
+	fu_mbim_qdu_updater_set_update_session (self->mbim_device, &ctx);
+	g_main_loop_run (mainloop);
+
+	if (ctx.error != NULL) {
+		g_propagate_error (error, ctx.error);
+		return NULL;
+	}
+
```

Comment:
`ctx.chunks` and `ctx.cancellable` would be all leaking here

Reply:
For chunks I think it's fine, yes.
For cancellable, just remove the cancellable and pass NULL to the requests, you don't need the cancellable.

Reply:
Yes, exactly :)